compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12845,6 +12845,151 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
+      //unsignedMin is always 0
+      bool canTransformAdd = isAddOp
+                             && !(oldUConst2 < oldUConst1);
+      //sub op must handle overflow
+      //x-a>b --opt--> x>b+a CASE OF OVERFLOW IF b+a>unsignedMax
+      bool canTransformSub = isSubOp
+                             && !(oldUConst2 > unsignedMax - oldUConst1);
+
+      if (!canTransformOp(canTransformAdd, canTransformSub, node, s))
+         return NULL;
+
+      uint64_t newUConst = 0;
+      newUConst = isAddOp ? (oldUConst2 - oldUConst1) : (oldUConst2 + oldUConst1);
+      // It's not recommended to mutate the value of an existing constant node by
+      // calling constNode->setConstValue()
+      // Hence, create and set new node here,
+      TR::Node *newConstNode = TR::Node::create(secondChild, secondChild->getOpCodeValue(), 0);
+      newConstNode->setUnsignedLongInt(newUConst);
+      return newConstNode;
+
+      }
+   else
+      {
+      int64_t signedMax = 0, signedMin = 0, oldConst1 = 0, oldConst2 = 0;
+      switch (DataType)
+         {
+         case TR::Int8:
+            oldConst1 = opNode->getSecondChild()->getByte();
+            oldConst2 = secondChild->getByte();
+            signedMax = TR::getMaxSigned<TR::Int8>();
+            signedMin = TR::getMinSigned<TR::Int8>();
+            break;
+         case TR::Int16:
+            oldConst1 = opNode->getSecondChild()->getShortInt();
+            oldConst2 = secondChild->getShortInt();
+            signedMax = TR::getMaxSigned<TR::Int16>();
+            signedMin = TR::getMinSigned<TR::Int16>();
+            break;
+         case TR::Int32:
+            oldConst1 = opNode->getSecondChild()->getInt();
+            oldConst2 = secondChild->getInt();
+            signedMax = TR::getMaxSigned<TR::Int32>();
+            signedMin = TR::getMinSigned<TR::Int32>();
+            break;
+         case TR::Int64:
+            oldConst1 = opNode->getSecondChild()->getLongInt();
+            oldConst2 = secondChild->getLongInt();
+            signedMax = TR::getMaxSigned<TR::Int64>();
+            signedMin = TR::getMinSigned<TR::Int64>();
+            break;
+         default:
+            if (s->trace())
+               traceMsg(s->comp(),
+                        "\nEliminating add/sub under compare node n%dn failed due to opcode data type\n",
+                        node->getGlobalIndex());
+            TR_ASSERT_FATAL(false, "\nEliminating add/sub under compare node n%dn failed due to opcode data type\n",
+                            node->getGlobalIndex());
+
+            break;
+         }
+      //for signed Constants, overflow and underflow are possible for either add or sub operation
+      //and both cases must be handled for each operation.
+      bool canTransformAdd = isAddOp
+                             && !(oldConst1 > 0 && (oldConst2 < signedMin + oldConst1))
+                             && !(oldConst1 < 0 && (oldConst2 > signedMax + oldConst1));
```

Comment:
IMHO, putting these checks into their own functions would make the code much easier to read.